### PR TITLE
feat(ts): add configs that support TypeScript @W-16036296@

### PR DESCRIPTION
**Pull request by wjhsf**
Properly setting the right `parserOptions` to support TypeScript is tedious, so this PR adds new configs that do that for you. Each existing config now also has a `-ts` flavor that supports TypeScript. For example, to use the recommended rules in a TypeScript project, use the `recommended-ts` config.

### CODE REVIEW

File: package.json
Comment (wjhsf) [inline package.json:33]: This is changed to `~` and is *not* the latest version of the package because in v28.5.0, they depend on `@typescript-eslint/utils@^6 ||| ^7`. Because we don't depend on that package ourselves, it resolves to v7, and v7 requires eslint 8. We still support eslint v7, so we can't do that.

I considered bumping to latest and leveraging `resolutions`, but that always feels more tedious than it's worth.

**Review by ravijayaramappa — APPROVED** [review]

**Comment by ravijayaramappa** [comment]
Overall this LGTM, preferably wait for another set of eyes to approve.

**Comment by wjhsf** [comment]
> Overall this LGTM, preferably wait for another set of eyes to approve.
Discussed during PR review sync; got no objections.

**Pull request merged**
